### PR TITLE
Implement xs{max,min}cqp and move xs{max,min}[cj]dp to decodetree

### PR DESCRIPTION
**Pull request by vcoracolombo**

### CODE REVIEW

File: target/ppc/translate/vsx-impl.c.inc [hunk 1]
```diff
@@ -2185,6 +2181,49 @@ TRANS(XXBLENDVH, do_xxblendv, MO_16)
 TRANS(XXBLENDVW, do_xxblendv, MO_32)
 TRANS(XXBLENDVD, do_xxblendv, MO_64)
 
+static bool do_xsmaxmin(int t, int a, int b,
```
Comment (vcoracolombo) [inline target/ppc/translate/vsx-impl.c.inc:2184]: I am thinking if this is the best approach. Should I remove this function and move its body to do_xsmaxmincjdp/do_xsmaxmincqp, changing do_xsmaxmincqp to use gen_avr_ptr?

File: target/ppc/translate/vsx-impl.c.inc [hunk 2]
```diff
@@ -2185,6 +2181,49 @@ TRANS(XXBLENDVH, do_xxblendv, MO_16)
...
+    ptr_b = gen_vsr_ptr(b);
+
+    helper(cpu_env, ptr_t, ptr_a, ptr_b);
+
+    tcg_temp_free_ptr(ptr_t);
+    tcg_temp_free_ptr(ptr_a);
+    tcg_temp_free_ptr(ptr_b);
+
+    return true;
+}
```
Comment (vcoracolombo) [inline target/ppc/translate/vsx-impl.c.inc:2200]: I am thinking if this is the best approach. Should I remove this function and move its body to do_xsmaxmincjdp/do_xsmaxmincqp, changing do_xsmaxmincqp to use gen_avr_ptr?

Comment (mferst) [reply target/ppc/translate/vsx-impl.c.inc:2200]: I see two other options:
1 - Receive three `TCGv_ptr` in this method, call `gen_vsr_ptr` in `do_xsmaxmincjdp` and `gen_avr_ptr` in `do_xsmaxmincqp`;
2 - Receive a `TCGv_ptr (*gen_ptr)(int reg)` parameter here.

The first splits resource acquisition and release, but things are quite localized here, it shouldn't be a problem. The second seems overkill for such a simple translation code...

File: target/ppc/insn32.decode [hunk 3]
```diff
@@ -128,6 +128,10 @@
 %xx2_xb         1:1 11:5
 @XX2            ...... ..... ... uim:2 ..... ......... ..       &XX2 xt=%xx2_xt xb=%xx2_xb
 
+&XX3            xt xa xb
+%xx3_xa         2:1 16:5
```
Comment (mferst) [inline target/ppc/insn32.decode:132]: These fields aren't XX2/XX3 only, so we are calling them `%xx_x[tabc]` in other commits, e.g.: https://github.com/PPC64/qemu/pull/36/commits/60f567b9d2ad4714d5e61ee496a48c9e2a47a566

File: target/ppc/insn32.decode [hunk 4]
```diff
@@ -427,3 +431,10 @@ XXSPLTW         111100 ..... ---.. ..... 010100100 . .  @XX2
 ## VSX Vector Load Special Value Instruction
 
 LXVKQ           111100 ..... 11111 ..... 0101101000 .   @X_uim5
+
+## VSX Comparison Instructions
+
+XSMAXCDP	111100 ..... ..... ..... 10000000 ...   @XX3
+XSMINCDP	111100 ..... ..... ..... 10001000 ...   @XX3
+XSMAXJDP	111100 ..... ..... ..... 10010000 ...   @XX3
+XSMINJDP	111100 ..... ..... ..... 10011000 ...   @XX3
```
Comment (mferst) [inline target/ppc/insn32.decode]: tab -> spaces
```suggestion
XSMAXCDP        111100 ..... ..... ..... 10000000 ...   @XX3
XSMINCDP        111100 ..... ..... ..... 10001000 ...   @XX3
XSMAXJDP        111100 ..... ..... ..... 10010000 ...   @XX3
XSMINJDP        111100 ..... ..... ..... 10011000 ...   @XX3
```

Comment (mferst) [reply target/ppc/insn32.decode]: doh, fixed in https://github.com/PPC64/qemu/pull/51/commits/7b280c144de38c0ab2cc8d0b6bf4d6cf42647d8b
Nothing to see here :)

File: target/ppc/insn32.decode [hunk 5]
```diff
@@ -427,3 +431,12 @@ XXSPLTW         111100 ..... ---.. ..... 010100100 . .  @XX2
...
 LXVKQ           111100 ..... 11111 ..... 0101101000 .   @X_uim5
+
+## VSX Comparison Instructions
+
+XSMAXCDP        111100 ..... ..... ..... 10000000 ...   @XX3
+XSMINCDP        111100 ..... ..... ..... 10001000 ...   @XX3
+XSMAXJDP        111100 ..... ..... ..... 10010000 ...   @XX3
+XSMINJDP        111100 ..... ..... ..... 10011000 ...   @XX3
+XSMAXCQP        111111 ..... ..... ..... 1010100100 -   @VX
+XSMINCQP        111111 ..... ..... ..... 1011100100 -   @VX
```
[the remark this answers is not in the document]
Comment (vcoracolombo) [reply target/ppc/insn32.decode]: Fixed, thanks

File: target/ppc/helper.h [hunk 6]
```diff
@@ -392,10 +392,12 @@ DEF_HELPER_4(xscmpoqp, void, env, i32, vsr, vsr)
...
-DEF_HELPER_5(xsmaxcdp, void, env, i32, vsr, vsr, vsr)
-DEF_HELPER_5(xsmincdp, void, env, i32, vsr, vsr, vsr)
-DEF_HELPER_5(xsmaxjdp, void, env, i32, vsr, vsr, vsr)
-DEF_HELPER_5(xsminjdp, void, env, i32, vsr, vsr, vsr)
+DEF_HELPER_4(xsmaxcdp, void, env, vsr, vsr, vsr)
+DEF_HELPER_4(xsmincdp, void, env, vsr, vsr, vsr)
+DEF_HELPER_4(xsmaxcqp, void, env, vsr, vsr, vsr)
+DEF_HELPER_4(xsmincqp, void, env, vsr, vsr, vsr)
+DEF_HELPER_4(xsmaxjdp, void, env, vsr, vsr, vsr)
+DEF_HELPER_4(xsminjdp, void, env, vsr, vsr, vsr)
```
[the remark this answers is not in the document]
Comment (vcoracolombo) [reply target/ppc/helper.h]: Fixed, thanks

**Review by mferst — APPROVED** [review]

**Comment by mferst** [comment]
Merged in upstream: https://lists.gnu.org/archive/html/qemu-ppc/2022-03/msg00002.html

**Pull request closed without merging**